### PR TITLE
PR for qbank comment

### PR DESCRIPTION
**Pull request by safatshahin**

### CODE REVIEW

**Review by guillogo — CHANGES_REQUESTED** [review]
Hi @safatshahin 

The PR has been reviewed. 

**Passed:**

- Feature test: OK
- User tests: Ok
- Capabilities test: OK (just thinking.. What do you think of the idea of having a deletecomment capability? to handle the possible loss of information)
- Unit test: OK
- Behat test: (has been added) running @core_question -> OK
- Code checker OK

**To fix:**
- Backup and restore: Getting this exception when try to restore a course that has comments on a question.
```
Exception - Call to undefined method restore_course_task::get_comment_mapping_itemname()
```
- Code: Some comments in the PR to check.


Thank you.
Guillermo

File: question/bank/comment/lib.php [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    if ($args->commentarea != 'core_question' && $args->component != 'qbank_comment') {
+        throw new comment_exception('core_question');
+    }
+    return $comments;
+}
+
+/**
+ * Comment content for callbacks.
+ *
+ * @param stdClass $question
```
[the remark this answers is not in the document]
Comment (safatshahin) [reply question/bank/comment/lib.php:75]: Actually, question_definition is a question object, and this method is used for a callback to the preview or any other plugin, the fragment is the only method using question definition from the question engine.

File: question/bank/comment/classes/comment_count_column.php [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+    public function get_name(): string {
+        return 'commentcount';
+    }
+
+    protected function get_title(): string {
+        return get_string('commentplural', 'qbank_comment');
+    }
+
+    protected function display_content($question, $rowclasses): void {
```
[the remark this answers is not in the document]
Comment (safatshahin) [reply question/bank/comment/classes/comment_count_column.php:50]: it was a oneline <a> tag without any attribute, that's y didn't create a template for a line

**Comment by safatshahin** [comment]
Hi @guillogo 
Thank you for your review of this. I have added some major codes to implement backup and restore in the core_backup which implements the backup and restore of comments for questions in course, quiz, or any other context system or categories. It also makes sure the question comment backup and restore is done with the question, not the regular comments XML structure.
Comment API has its own moodle/comment:delete capability which it checks in the interactive interface before triggering the delete from the js. It's a global feature from the API and I don't think there is anything can be done if we continue to use ajax based comments API. Hopefully, the comments API rebuild will do something good for this (there is a project that will start soon on that).
Please give it another go and let me know of any changes.
Cheers!

**Comment by safatshahin** [comment]
Hi @guillogo 
I added the backup and restore with fixed issues in a new implementation, added deletion, behat tests and css fix.
Thanks

**Review by guillogo — APPROVED** [review]
Approved.

Tested all scenarios including backup and restore in a different ways.

**Pull request merged**